### PR TITLE
k8s reporter, Extend collector to include network commands

### PR DESCRIPTION
**Pull request by oshoval**
Extend k8s reporter to collect networking commands upon assertion 
or when triggered by `make dump`.

Add commands for
1. Nodes (see `executeNodeCommands`)
2. Virt-launchers (see `executeVirtLauncherCommands`)
3. VMIs (see `executeVMICommands`)

This PR also fixes re-login of alpine.
 
Signed-off-by: Or Shoval <oshoval@redhat.com>

```release-note
None
```

### CODE REVIEW

File: tests/reporter/kubernetes.go [hunk 1]
```diff
@@ -918,3 +1024,157 @@ func writeStringToFile(filePath string, data string) error {
 	_, err = f.WriteString(data)
 	return err
 }
+
+func getVmiType(vmi v12.VirtualMachineInstance) (string, error) {
+	for _, volume := range vmi.Spec.Volumes {
+		if volume.VolumeSource.ContainerDisk == nil {
+			continue
```
Comment (alonSadan) [inline tests/reporter/kubernetes.go:1031]: I think it would help with debugging if you will return a different error here
```
return "", fmt.Errorf("containerDisk is nil for vmi  %s", vmi.ObjectMeta.Name)
```

File: tests/reporter/kubernetes.go [hunk 2]
```diff
@@ -918,3 +1024,157 @@ func writeStringToFile(filePath string, data string) error {
...
+		if strings.Contains(image, "fedora") {
+			return "fedora", nil
+		} else if strings.Contains(image, "cirros") {
+			return "cirros", nil
+		} else if strings.Contains(image, "alpine") {
+			return "alpine", nil
+		}
+	}
+
+	return "", fmt.Errorf("unknown type, vmi %s", vmi.ObjectMeta.Name)
```
Comment (alonSadan) [inline tests/reporter/kubernetes.go:1044]: If you choose to return if containerDisk is nil, you can also return the unknown type here:
```
return "", fmt.Errorf("unknown type %s for vmi %s", volume.VolumeSource.ContainerDisk.Image, vmi.ObjectMeta.Name) 
```

File: tests/reporter/kubernetes.go [hunk 3]
```diff
@@ -918,3 +1024,157 @@ func writeStringToFile(filePath string, data string) error {
...
+			return console.LoginToCirros(&vmi)
+		case "alpine":
+			return console.LoginToAlpine(&vmi)
+		}
+	}
+
+	return err
+}
+
+func checkPodComputeContainer(pod v1.Pod) bool {
```
Comment (alonSadan) [inline tests/reporter/kubernetes.go:1067]: nit: maybe change to `isPodComputeContainer`

**Review by alonSadan — COMMENTED** [review]
Thanks Or. Really nice. some minor comments

File: tests/reporter/kubernetes.go (hunk 3 above)
Comment (oshoval) [reply tests/reporter/kubernetes.go:1067]: but the pod is not a compute container, it might have a compute container, this is what the function checks

File: tests/reporter/kubernetes.go (hunk 2 above)
Comment (oshoval) [reply tests/reporter/kubernetes.go:1044]: It might have several ContainerDisk volumes IIUC,
if so i can't print them

File: tests/reporter/kubernetes.go (hunk 1 above)
Comment (oshoval) [reply tests/reporter/kubernetes.go:1031]: Some volumes don't have ContainerDisk and its ok, there might be couple volumes, we don't want to print all the non ContainerDisk ones, but we do `continue`

File: tests/reporter/kubernetes.go (hunk 3 above)
Comment (alonSadan) [reply tests/reporter/kubernetes.go:1067]: exactly, boolean functions like this one, are a lot of times called `isSomething`  or `areSomething`.
for exapmle check out the [first result](https://levelup.gitconnected.com/knot-of-nots-avoiding-negative-names-for-boolean-methods-641896a94a42#:~:text=The%20usual%20convention%20to%20name,exists%2C%20you%20would%20say%20user.) in google for "boolean function naming convention".

there is no known convention for `check Something` AFAIK, and it could mean many things.

**Comment by oshoval** [comment]
Thanks Alon,
addressed comments

File: tests/reporter/kubernetes.go [hunk 4]
```diff
@@ -134,6 +142,10 @@ func (r *KubernetesReporter) Dump(duration time.Duration) {
 	r.logDomainXMLs(virtCli, vmis)
 
 	r.logSRIOVInfo(virtCli)
+
+	r.logNodeCommands(virtCli)
```
Comment (alonSadan) [inline tests/reporter/kubernetes.go:146]: Is there anything else other than network info that is retrieved from `logNodeCommands`, `logVirtLauncherCommands`, and `logVMICommands` ?
Perhaps it would be better to put it in a "network context" and call those functions :
`logNodeNetworkInfo`, `logVirtLauncherNetworkInfo`, and `logVMINetworkInfo`?
or, the same but with 'NetworkState` instead of `Commands`?

File: tests/reporter/kubernetes.go [hunk 5]
```diff
@@ -337,6 +349,100 @@ func (r *KubernetesReporter) logAuditLogs(virtCli kubecli.KubevirtClient, logsdi
...
+				fmt.Fprintf(os.Stderr, "failed get vmi type: %v\n", err)
+				continue
+			}
+
+			if err := prepareVmiConsole(vmi, vmiType); err != nil {
+				fmt.Fprintf(os.Stderr, "failed login to vmi: %v\n", err)
+				continue
+			}
+
+			r.executeVMICommands(vmi, logsdir, vmiType)
```
Comment (alonSadan) [inline tests/reporter/kubernetes.go:389]: same as here https://github.com/kubevirt/kubevirt/pull/6116/files#r677171619
maybe, since it looks like this is only networking commands, rename this function to show this is only network related commands.

File: tests/reporter/kubernetes.go [hunk 6]
```diff
@@ -337,6 +349,100 @@ func (r *KubernetesReporter) logAuditLogs(virtCli kubecli.KubevirtClient, logsdi
...
+		return
+	}
+
+	for _, pod := range pods.Items {
+		if pod.Status.Phase != "Running" {
+			fmt.Fprintf(os.Stderr, "skipping node's pod %s, phase is not Running\n", pod.ObjectMeta.Name)
+			continue
+		}
+
+		r.executeNodeCommands(virtCli, logsdir, pod)
```
Comment (alonSadan) [inline tests/reporter/kubernetes.go:442]: ditto

File: tests/reporter/kubernetes.go (hunk 4 above)
Comment (oshoval) [reply tests/reporter/kubernetes.go:146]: Answering on all the comments at once:
its almost true that they run only network commands but
1. dmesg is not a network command, so maybe not perfectly true.
2. we might want to add them to the future commands which are not network

having said that, it dumps the info in network folder

considered that, wdyt we should do ? (about all comments, since there are two groups ,this one, and the inner functions of the other comments)

File: tests/reporter/kubernetes.go [hunk 7]
```diff
@@ -337,6 +349,100 @@ func (r *KubernetesReporter) logAuditLogs(virtCli kubecli.KubevirtClient, logsdi
...
+			fmt.Fprintf(os.Stderr, "skipping pod %s, phase is not Running\n", pod.ObjectMeta.Name)
+			continue
+		}
+
+		if found := podHasComputeContainer(pod); found != true {
+			fmt.Fprintf(os.Stderr, "could not find compute container for pod %s\n", pod.ObjectMeta.Name)
+			continue
+		}
+
+		r.executeVirtLauncherCommands(virtCli, logsdir, pod)
```
Comment (alonSadan) [inline tests/reporter/kubernetes.go:419]: ditto

File: tests/reporter/kubernetes.go (hunk 4 above)
Comment (alonSadan) [reply tests/reporter/kubernetes.go:146]: I see your point. Fine by me to leave it as is

**Review by alonSadan — APPROVED** [review]
Thank you:)

**Comment by oshoval** [comment]
Thanks Alon
@AlonaKaplan can you take a look please?

File: tests/reporter/kubernetes.go [hunk 8]
```diff
@@ -918,3 +1024,157 @@ func writeStringToFile(filePath string, data string) error {
...
+	cmds := []commands{
+		{command: "ip address", fileNameSuffix: "ipaddress"},
+		{command: "ip link", fileNameSuffix: "iplink"},
+		{command: "ip route show table all", fileNameSuffix: "iproute"},
+		{command: "ip neigh show", fileNameSuffix: "ipneigh"},
+		{command: "bridge -j vlan show", fileNameSuffix: "brvlan"},
+		{command: "bridge fdb", fileNameSuffix: "brfdb"},
+		{command: "lspci", fileNameSuffix: "lspci"},
+		{command: "env", fileNameSuffix: "env"},
+		{command: "nft list ruleset", fileNameSuffix: "nftlist"},
```
[the remark this answers is not in the document]
Comment (oshoval) [reply tests/reporter/kubernetes.go:1106]: The binary is there, but output is empty indeed, i think youright
is there a simple code example please on how to access the launcher namespace from virt-handler ?

File: tests/reporter/kubernetes.go [hunk 9]
```diff
@@ -918,3 +1024,157 @@ func writeStringToFile(filePath string, data string) error {
...
+		} else if strings.Contains(image, "alpine") {
+			return "alpine", nil
+		}
+	}
+
+	return "", fmt.Errorf("unknown type, vmi %s", vmi.ObjectMeta.Name)
+}
+
+func prepareVmiConsole(vmi v12.VirtualMachineInstance, vmiType string) error {
+	err := console.SafeExpectBatch(&vmi, []expect.Batcher{
```
[the remark this answers is not in the document]
Comment (oshoval) [reply tests/reporter/kubernetes.go:1048]: the login fails if its already logged
for example if the test did login already and then asserted

File: tests/reporter/kubernetes.go [hunk 10]
```diff
@@ -337,6 +349,100 @@ func (r *KubernetesReporter) logAuditLogs(virtCli kubecli.KubevirtClient, logsdi
...
+
+	logsdir := filepath.Join(r.artifactsDir, "network", "vmis")
+	if err := os.MkdirAll(logsdir, 0777); err != nil {
+		fmt.Fprintf(os.Stderr, "failed to create directory: %v\n", err)
+		return
+	}
+
+	vmiNamespaces := tests.TestNamespaces
+	if isExternal {
+		vmiNamespaces = []string{v1.NamespaceAll}
```
[the remark this answers is not in the document]
Comment (oshoval) [reply tests/reporter/kubernetes.go:362]: Consoling a VM is disruptive, it disconnected the previous logged instances.
If we run in parallel and collect all VMs, we will hurt healthy tests during their session (or other ongoing assertions),

when running from e2e assertion, we can use `TestNamespaces` in order to touch only the namespaces
where tests should run their VMs, and each parallel instance of e2e test has already its own `TestNamespaces`

If we run `make dump` manually, we do want to collect all

File: tests/reporter/kubernetes.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply tests/reporter/kubernetes.go:1106]: That's how it is [done](https://github.com/kubevirt/kubevirt/blob/65656ff9141ba830e03be5cbdf32439a943ea68a/pkg/virt-handler/isolation/detector.go#L77) in the virt-handler code. I guess you can simulate the same behaviour.
Another option is to do on the virt-handler -
`ps -ef | grep " --name vmi-masquerade " | grep /usr/bin/virt-launcher | head -n1 | cut -d" " -f7` to find the launcher pid and then enter the pid using nsenter.

File: tests/reporter/kubernetes.go [hunk 11]
```diff
@@ -918,3 +1024,157 @@ func writeStringToFile(filePath string, data string) error {
...
+	cmds := []commands{
+		{command: networkPrefix + "ip address", fileNameSuffix: "ipaddress"},
+		{command: networkPrefix + "ip link", fileNameSuffix: "iplink"},
+		{command: networkPrefix + "ip route show table all", fileNameSuffix: "iproute"},
+		{command: networkPrefix + "ip neigh show", fileNameSuffix: "ipneigh"},
+		{command: networkPrefix + "bridge -j vlan show", fileNameSuffix: "brvlan"},
+		{command: networkPrefix + "bridge fdb", fileNameSuffix: "brfdb"},
+		{command: networkPrefix + "nft list ruleset", fileNameSuffix: "nftlist"},
+
+		{command: hostPrefix + "/usr/bin/" + networkPrefix + "/usr/sbin/iptables --list", fileNameSuffix: "iptables"},
```
Comment (AlonaKaplan) [inline tests/reporter/kubernetes.go:1090]: I think the interesting part in the iptables output is the counters. I would add the `-v` option.

**Comment by oshoval** [comment]
addressed comments

**Comment by oshoval** [comment]
we can see nftlist on kind here for example
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6116/pull-kubevirt-e2e-kind-1.19-sriov/1420712515519647744

(network/pods folder)

**Review by alonSadan — APPROVED** [review]

File: tests/reporter/kubernetes.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply tests/reporter/kubernetes.go:1093]: Our login methods should support re-login. If one of the three doesn't support it, it should be fixed.

[the remark this answers is not in the document]
Comment (oshoval) [reply tests/reporter/kubernetes.go]: Done

**Comment by oshoval** [comment]
addressed comments

**Comment by oshoval** [comment]
/retest 

doesnt seem related but lets make sure
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6116/pull-kubevirt-e2e-k8s-1.19-sig-storage/1422158509151096832

**Comment by oshoval** [comment]
another weird failures
retesting but noting down 
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6116/pull-kubevirt-e2e-k8s-1.20-sig-compute-nonroot/1422158508349984768

/retest

File: tests/reporter/kubernetes.go [hunk 12]
```diff
@@ -337,6 +350,133 @@ func (r *KubernetesReporter) logAuditLogs(virtCli kubecli.KubevirtClient, logsdi
...
+		r.executeVirtLauncherCommands(virtCli, logsdir, pod)
+	}
+
+	for _, virtHandlerPod := range virtHandlerPods.Items {
+		if virtHandlerPod.Status.Phase != "Running" {
+			fmt.Fprintf(os.Stderr, "skipping virt-handler %s, phase is not Running\n", virtHandlerPod.ObjectMeta.Name)
+			continue
+		}
+
+		processList, err := getProcessList(virtCli, &virtHandlerPod)
```
Comment (AlonaKaplan) [inline tests/reporter/kubernetes.go:429]: Instead of the suggested code (that I believe is very complicated) I would create a map node_name -> pair list(virt_launcher_pod_name, vm_name).
Then for each virt-handler I would execute for each vmi `pgrep -f "name vmi_name.*no-fork"`, it should give you the pid of the virt-launcher pod.

What do you think?

File: tests/reporter/kubernetes.go [hunk 13]
```diff
@@ -337,6 +350,133 @@ func (r *KubernetesReporter) logAuditLogs(virtCli kubecli.KubevirtClient, logsdi
...
+		if err != nil {
+			continue
+		}
+
+		lines := strings.Split(processList, "\n")
+		for _, line := range lines {
+			if strings.Contains(line, "no-fork") {
+				ppid, uid, namespace := parseProcessLine(line)
```
Comment (AlonaKaplan) [inline tests/reporter/kubernetes.go:437]: Why the ppid is required isn't the pid enough. Both the for and the regular process have the same network name space.

File: tests/console/login.go [hunk 14]
```diff
@@ -77,7 +77,20 @@ func LoginToAlpine(vmi *v1.VirtualMachineInstance) error {
 	}
 	defer expecter.Close()
 
+	// Do not login, if we already logged in
 	b := append([]expect.Batcher{
+		&expect.BSnd{S: "\n"},
+		&expect.BExp{R: PromptExpression},
+		&expect.BSnd{S: "true\n"},
```
Comment (AlonaKaplan) [inline tests/console/login.go:84]: What is the added value of sending `true`?

File: tests/console/login.go [hunk 15]
```diff
@@ -117,10 +130,11 @@ func LoginToFedora(vmi *v1.VirtualMachineInstance) error {
 	b := append([]expect.Batcher{
 		&expect.BSnd{S: "\n"},
 		&expect.BExp{R: PromptExpression},
+		&expect.BSnd{S: "true\n"},
```
Comment (AlonaKaplan) [inline tests/console/login.go:133]: I don't understand why the `true` command is needed here. Couldn't understand the explanation in the commit message as well.

File: tests/reporter/kubernetes.go (hunk 12 above)
Comment (oshoval) [reply tests/reporter/kubernetes.go:429]: pgrep didn't work well via console
we also need the uid, in order to find the virt-launcher name (in order to save to the right file name), so parsing is needed anyhow

side note: the right thing to do is use the uid instead of vmi_name + (namespace which is needed)
as we are doing here `GetRunningPodByLabel`

File: tests/reporter/kubernetes.go (hunk 13 above)
Comment (oshoval) [reply tests/reporter/kubernetes.go:437]: pid can be used i believe as well indeed, just chosen the original process because it felt more "natual"
i will try the pid

File: tests/reporter/kubernetes.go [hunk 16]
```diff
@@ -918,3 +1069,240 @@ func writeStringToFile(filePath string, data string) error {
...
+		fmt.Fprintf(os.Stderr, "failed to write %s %s output: %v\n", target, "nftlist", err)
+		return
+	}
+}
+
+func getProcessList(virtCli kubecli.KubevirtClient, virtHandlerPod *v1.Pod) (string, error) {
+	command := []string{
+		"/bin/bash",
+		"-c",
+		"trap '' URG && /usr/bin/ps -ef",
```
[the remark this answers is not in the document]
Comment (oshoval) [reply tests/reporter/kubernetes.go:1244]: ps might panic due to URG signal

```
Signal 23 (URG) caught by ps (3.3.16).
/usr/bin/ps:ps/display.c:66: please report this bug
```

see here please
https://github.com/kubevirt/kubevirt/pull/5846/commits/3b16366b0af8f162701460cce57f9b0624358968

File: tests/reporter/kubernetes.go [hunk 17]
```diff
@@ -337,6 +350,133 @@ func (r *KubernetesReporter) logAuditLogs(virtCli kubecli.KubevirtClient, logsdi
...
+			fmt.Fprintf(os.Stderr, "skipping virt-handler %s, phase is not Running\n", virtHandlerPod.ObjectMeta.Name)
+			continue
+		}
+
+		processList, err := getProcessList(virtCli, &virtHandlerPod)
+		if err != nil {
+			continue
+		}
+
+		lines := strings.Split(processList, "\n")
```
[the remark this answers is not in the document]
Comment (oshoval) [reply tests/reporter/kubernetes.go:434]: This is what i tried to do with `parseProcessLine` (line 437) and line (429), so please elaborate
lets discuss offline about all the points here if possible (this one and https://github.com/kubevirt/kubevirt/pull/6116/files#r681184802)
since they are related to each other and have many small details

File: tests/console/login.go (hunk 15 above)
Comment (oshoval) [reply tests/console/login.go:133]: Lets say one sends command that fails, but its output isn't consumed 
then the relogin will find the PromptExpression but the echo $? will failed and the flow will continue the login process below (it will find the prompt there, so it won't send user:pass again, but it will reconfigure again the shell, and the whole thing is slower)

site node - this can be improved on other PR to match the full prompt (localhost|%s) login: `, vmi.Name)
instead of PromptExpression
because PromptExpression can be even matched during boot when the machine isn't actually logged
see here for example test 1705 when i totally removed the 
```
		&expect.BSnd{S: "echo $?\n"},
		&expect.BExp{R: RetValue("0")},
```
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6116/pull-kubevirt-e2e-k8s-1.20-sig-network/1422158507934748672
Then it failed because it didn't need more condition after it matched the `# `
the `echo $?` makes it better but still not perfect but not need to be part of this PR

even adding the true can be done on other PR if needed basically, but it does improve the login in case the last command fail and its output was not consumed (for example timeout, or if ran manually and then ran `make dump`)

File: tests/console/login.go (hunk 14 above)
Comment (oshoval) [reply tests/console/login.go:84]: same as https://github.com/kubevirt/kubevirt/pull/6116/files#r681193075

**Comment by oshoval** [comment]
addressed the comments that about the collector (not yet about the login commit)

File: tests/reporter/kubernetes.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (oshoval) [reply tests/reporter/kubernetes.go]: Note that the drawback of it (it was existing already on fedora and cirros, i just added to alpine)
is that it will consume several seconds in the tests themself for every first login

**Comment by oshoval** [comment]
Improved the login
Please see 
https://github.com/kubevirt/kubevirt/compare/a42be5ebf59b64c9298ab94599eb078a1174b640..bbc94be34e5075fe1cae400e9cd80078a0cc8da4

We might want to do "sudo su" again on the relogin block ?
(in case a user connected but without sudo)
currently the commands that we are collecting on the VM doesn't need sudo, so currently it 
shouldn't be an issue

**Comment by oshoval** [comment]
Removed function i dont need anymore

File: tests/reporter/kubernetes.go [hunk 18]
```diff
@@ -337,6 +352,142 @@ func (r *KubernetesReporter) logAuditLogs(virtCli kubecli.KubevirtClient, logsdi
...
+		if virtHandlerPod.Status.Phase != "Running" {
+			fmt.Fprintf(os.Stderr, "skipping virt-handler %s, phase is not Running\n", virtHandlerPod.ObjectMeta.Name)
+			continue
+		}
+
+		nodeMap[virtHandlerPod.Spec.NodeName] = virtHandlerPod
+	}
+
+	vmis, err := virtCli.VirtualMachineInstance(v1.NamespaceAll).List(&metav1.ListOptions{})
+	if err != nil {
```
Comment (AlonaKaplan) [inline tests/reporter/kubernetes.go:415]: It looks great overall, just a small question. Maybe it is enough to query for the list of virt-launcher pods and no need to query the vmi list. Since the virt-launcher pod contains the name of the vmi?

File: tests/console/login.go [hunk 19]
```diff
@@ -116,11 +126,9 @@ func LoginToFedora(vmi *v1.VirtualMachineInstance) error {
 	// Do not login, if we already logged in
 	b := append([]expect.Batcher{
 		&expect.BSnd{S: "\n"},
-		&expect.BExp{R: PromptExpression},
-		&expect.BSnd{S: "echo $?\n"},
-		&expect.BExp{R: RetValue("0")},
+		&expect.BExp{R: fmt.Sprintf(`(\[fedora@(localhost|%s) ~\]\$ |\[root@(localhost|%s) fedora\]\# )`, vmi.Name, vmi.Name)},
```
Comment (AlonaKaplan) [inline tests/console/login.go:129]: I guess cirros needs a similar fix. Unless it blocks this PR I believe the login part deserves a separate PR.

File: tests/reporter/kubernetes.go (hunk 18 above)
Comment (oshoval) [reply tests/reporter/kubernetes.go:415]: I think you are right (you mean the name of the virt-launcher i guess which is what we need)
trying

**Comment by oshoval** [comment]
Rebased

**Comment by oshoval** [comment]
Addressed comments
added this as well
https://github.com/kubevirt/kubevirt/compare/b4a7fbd35ed63260c2da5190a43e7f1d91c0eb88..9347fe3db7298502f02f1a15e085ddf1c8032944#diff-dc90a95f97ee6623eb2cd8b29fd07ed28cda2fa3cc87aa75127900ec76da5142R80

File: tests/console/login.go (hunk 19 above)
Comment (oshoval) [reply tests/console/login.go]: https://github.com/kubevirt/kubevirt/pull/6181

Comment (oshoval) [reply tests/console/login.go]: On Cirros there is no issue, because the prompt of both sudo and non sudo are just `$` and this is what it waits

if `$` was printed in the middle of the boot it was an issue, but then we would saw it before already

File: tests/console/login.go [hunk 20]
```diff
@@ -77,7 +77,22 @@ func LoginToAlpine(vmi *v1.VirtualMachineInstance) error {
 	}
 	defer expecter.Close()
 
+	err = expecter.Send("\n")
```
Comment (AlonaKaplan) [inline tests/console/login.go:80]: Why lines 80-84 are needed? isn't the batcher enough?

Comment (oshoval) [reply tests/console/login.go:80]: Its basically enough, but those are little faster, sanity that we can even send,
saw it as well on fedora and Cirros so added here

**Comment by oshoval** [comment]
Thanks Alona
@alonSadan can you take a look please?

**Review by alonSadan — APPROVED** [review]

**Comment by oshoval** [comment]
Thanks Alon

/retest
